File: app/src/main/java/com/travlog/android/apps/ui/activities/NoteDetailsActivity.java
package com.travlog.android.apps.ui.activities;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.AppCompatTextView;
import android.support.v7.widget.Toolbar;
import android.util.Pair;
import android.view.Menu;
import android.view.MenuItem;

import com.travlog.android.apps.R;
import com.travlog.android.apps.libs.BaseActivity;
import com.travlog.android.apps.libs.qualifiers.RequiresActivityViewModel;
import com.travlog.android.apps.models.Note;
import com.travlog.android.apps.viewmodels.NoteDetailsViewModel;

import butterknife.BindView;
import io.reactivex.android.schedulers.AndroidSchedulers;

import static com.travlog.android.apps.libs.utils.TransitionUtilsKt.slideInFromLeft;
import static com.travlog.android.apps.ui.IntentKey.NOTE;

@RequiresActivityViewModel(NoteDetailsViewModel.class)
public class NoteDetailsActivity extends BaseActivity<NoteDetailsViewModel> {

    @BindView(R.id.toolbar)
    Toolbar toolbar;
    @BindView(R.id.title)
    AppCompatTextView title;

    @SuppressLint("CheckResult")
    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.a_note_details);
        setSupportActionBar(toolbar);
        setDisplayHomeAsUpEnabled(true);

        viewModel.outputs.setTitleText()
                .compose(bindToLifecycle())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::setTitleText);

        viewModel.outputs.showEditNoteActivity()
                .compose(bindToLifecycle())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::showEditNoteActivity);

        addDisposable(
                viewModel.outputs.back()
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(this::back));
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_note_details, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        final int id = item.getItemId();

        switch (id) {
            case R.id.menu_edit:
                viewModel.inputs.editClick();
                return true;
            case R.id.menu_delete:
                viewModel.inputs.deleteClick();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    private void setTitleText(final @NonNull String title) {
        this.title.setText(title);
    }

    private void showEditNoteActivity(final @NonNull Note note) {
        final Intent intent = new Intent(this, EditNoteActivity.class);
        intent.putExtra(NOTE, note);
        startActivityWithTransition(intent, R.anim.slide_in_right, R.anim.fade_out_slide_out_left);
    }

    @Nullable
    @Override
    protected Pair<Integer, Integer> exitTransition() {
        return slideInFromLeft();
    }
}

File: app/src/main/java/com/travlog/android/apps/ui/activities/NoteDetailsActivity.kt
/*
 * Copyright 2019 Travlog. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.travlog.android.apps.ui.activities

import android.content.Intent
import android.os.Bundle
import android.util.Pair
import android.view.Menu
import android.view.MenuItem
import androidx.databinding.DataBindingUtil
import com.jakewharton.rxbinding3.view.clicks
import com.travlog.android.apps.R
import com.travlog.android.apps.ViewModelFactory
import com.travlog.android.apps.databinding.ANoteDetailsBinding
import com.travlog.android.apps.getAppInjector
import com.travlog.android.apps.getViewModel
import com.travlog.android.apps.libs.BaseActivity
import com.travlog.android.apps.libs.utils.TransitionUtils.slideInFromLeft
import com.travlog.android.apps.models.Note
import com.travlog.android.apps.ui.IntentKey.NOTE_ID
import com.travlog.android.apps.ui.adapters.DestinationAdapter
import com.travlog.android.apps.viewmodels.NoteDetailsViewModel
import io.reactivex.android.schedulers.AndroidSchedulers
import kotlinx.android.synthetic.main.a_note_details.*
import javax.inject.Inject

class NoteDetailsActivity : BaseActivity<NoteDetailsViewModel>() {

    @Inject
    lateinit var viewModelFactory: ViewModelFactory

    private lateinit var destinationAdapter: DestinationAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        getAppInjector().inject(this)
        viewModel = getViewModel(viewModelFactory)

        super.onCreate(savedInstanceState)

        val binding = DataBindingUtil.setContentView<ANoteDetailsBinding>(this, R.layout.a_note_details)
                .apply {
                    setSupportActionBar(toolbar)
                    setDisplayHomeAsUpEnabled(true)
                }

        viewModel?.apply {
            destinationAdapter = DestinationAdapter(this)
                    .apply {
                        recycler_view.adapter = this
                    }

            delete_button.clicks()
                    .compose(bindToLifecycle())
                    .subscribe { inputs.deleteClick() }

            outputs.setNote()
                    .compose(bindToLifecycle())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(binding::setNote)

            outputs.updateDestinations()
                    .compose(bindToLifecycle())
                    .observeOn(AndroidSchedulers.mainThread())
                    .doOnNext { destinationAdapter.clearData() }
                    .subscribe { destinationAdapter.updateData(it) }

            outputs.showEditNote()
                    .compose(bindToLifecycle())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(::showEditNoteActivity)

            addDisposable(
                    outputs.finish()
                            .observeOn(AndroidSchedulers.mainThread())
                            .subscribe { back() }
            )
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_note_details, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        item.itemId.let {
            return when (it) {
                R.id.menu_edit -> {
                    viewModel?.inputs?.editClick()
                    true
                }
                else -> super.onOptionsItemSelected(item)
            }
        }
    }

    private fun showEditNoteActivity(note: Note) =
            Intent(this, EditNoteActivity::class.java).let {
                it.putExtra(NOTE_ID, note.id)
                startActivityWithTransition(it, R.anim.slide_in_right, R.anim.fade_out_slide_out_left)
            }

    override fun exitTransition(): Pair<Int, Int>? = slideInFromLeft()
}
